### PR TITLE
[CTRLRFC-1194][Hamamatsu] Add hability to manage Kind and Polarity output trigger options to Hamamatsu camera

### PR DESCRIPTION
**Pull request by goldensgui**
Add read and write methods to allow management of the KIND and POLARITY options related to equipment output triggers :

 - Read : 
   - void Hamamatsu::read_channel1Polarity(Tango::Attribute &attr)
   - void Hamamatsu::read_channel2Polarity(Tango::Attribute &attr)
   - void Hamamatsu::read_channel3Polarity(Tango::Attribute &attr)
   - void Hamamatsu::read_channel1Kind(Tango::Attribute &attr)
   - void Hamamatsu::read_channel2Kind(Tango::Attribute &attr)
   - void Hamamatsu::read_channel3Kind(Tango::Attribute &attr)

 - Write:
    - void Hamamatsu::write_channel1Polarity(Tango::WAttribute &attr)
    - void Hamamatsu::write_channel2Polarity(Tango::WAttribute &attr)
    - void Hamamatsu::write_channel3Polarity(Tango::WAttribute &attr)
    - void Hamamatsu::write_channel1Kind(Tango::WAttribute &attr)
    - void Hamamatsu::write_channel2Kind(Tango::WAttribute &attr)
    - void Hamamatsu::write_channel3Kind(Tango::WAttribute &attr)

 - Attributes created : 
     - attr_channel1Kind
     - attr_channel2Kind 
     - attr_channel3Kind
     - attr_channel1Polarity
     - attr_channel2Polarity
     - attr_channel3Polarity

### CODE REVIEW

File: specifics/Hamamatsu/HamamatsuStateMachine.cpp [hunk 1]
```diff
@@ -149,9 +149,93 @@ bool Hamamatsu::is_bottomViewExposureTime_allowed(Tango::AttReqType type)
...
+//
+//-----------------------------------------------------------------------------
+bool Hamamatsu::is_channel1Polarity_allowed(Tango::AttReqType type)
+{
+		//	End of Generated Code
+
+		//	Re-Start of Generated Code
+	return true;
+}
+//+----------------------------------------------------------------------------
```
Comment (soleil-ica) [inline specifics/Hamamatsu/HamamatsuStateMachine.cpp:166]: no State machine? at least INIT and FAULT, maybe RUNNING ?

File: specifics/Merlin/MerlinClass.cpp [hunk 2]
```diff
@@ -720,7 +720,7 @@ void MerlinClass::attribute_factory(vector<Tango::Attr *> &att_list)
 	//	Attribute : gain
 	gainAttrib	*gain = new gainAttrib();
 	Tango::UserDefaultAttrProp	gain_prop;
-    gain_prop.set_description("0 = SHGM\n1 = HGM\n2 = LGM\n3 = SLGM");
+    gain_prop.set_description("0 = SLGM \n1 = LGM\n2 = HGM\n3 = SHGM");
```
Comment (soleil-ica) [inline specifics/Merlin/MerlinClass.cpp:723]: I still don't understand why this commit appear there ... I hope that for the next commit, all the stuff from Hamamatsu will not be there... (did you do a pull before starting to work ?)

File: src/LimaDetector.cpp [hunk 3]
```diff
@@ -4664,7 +4664,7 @@ void LimaDetector::create_log_info_attributes(void)
 #endif
 
 		//- instanciate the log Adapter
-		yat4tango::YatLogAdapter::initialize(this);		
+		//yat4tango::YatLogAdapter::initialize(this);
```
Comment (soleil-ica) [inline src/LimaDetector.cpp:4667]: uncomment

File: src/LimaDetector.cpp [hunk 4]
```diff
@@ -4681,7 +4681,7 @@ void LimaDetector::delete_log_info_attributes()
 	yat4tango::DeviceInfo::release(this);
 
 	INFO_STREAM << "Remove the log-adapter." << endl;
-	yat4tango::YatLogAdapter::release();	
+	//yat4tango::YatLogAdapter::release();
```
Comment (soleil-ica) [inline src/LimaDetector.cpp:4684]: uncomment

**Review by soleil-ica — CHANGES_REQUESTED** [review]

**Review by soleil-ica — CHANGES_REQUESTED** [review]

File: doc/Hamamatsu/hamamatsu.rst [hunk 5]
```diff
@@ -32,12 +32,12 @@ fps                     R          Double     The last computed frame per second
...
 bottomViewExposureTime  R/W        Double     Exposure time for second W-VIEW.
-channel1Polarity		R/W		   UShort	  Polarity for channel 1 (NEGATIVE = 1, POSITIVE = 2).
-channel2Polarity		R/W		   UShort	  Polarity for channel 2 (NEGATIVE = 1, POSITIVE = 2).
-channel3Polarity		R/W		   UShort	  Polarity for channel 3 (NEGATIVE = 1, POSITIVE = 2).
-channel1Kind			R/W		   UShort	  Kind for channel 1 (LOW = 1, GLOBAL_EXPOSURE  = 1, PROGRAMMABLE = 2, TRIGGER_READY = 3, HIGH = 4).
-channel2Kind			R/W		   UShort	  Kind for channel 2 (LOW = 1, GLOBAL_EXPOSURE  = 1, PROGRAMMABLE = 2, TRIGGER_READY = 3, HIGH = 4).
-channel3Kind			R/W		   UShort	  Kind for channel 3 (LOW = 1, GLOBAL_EXPOSURE  = 1, PROGRAMMABLE = 2, TRIGGER_READY = 3, HIGH = 4).
+channel1Polarity	R/W	   UShort     Polarity for channel 1 (NEGATIVE = 1, POSITIVE = 2).
+channel2Polarity	R/W	   UShort     Polarity for channel 2 (NEGATIVE = 1, POSITIVE = 2).
+channel3Polarity	R/W	   UShort     Polarity for channel 3 (NEGATIVE = 1, POSITIVE = 2).
```
[the remark this answers is not in the document]
Comment (goldensgui) [reply doc/Hamamatsu/hamamatsu.rst:37]: SDK uses those values and we didn't want to interfere with its usage. Value can be corrected if needed/more convenient for user

[the remark this answers is not in the document]
Comment (soleil-ica) [reply doc/Hamamatsu/hamamatsu.rst:37]: we agreed to use: NEGATIVE = 0, POSITIVE = 1

**Review by soleil-ica — APPROVED** [review]

**Review by soleil-ica — APPROVED** [review]

**Review by soleil-ica — APPROVED** [review]

File: specifics/Hamamatsu/Hamamatsu.cpp [hunk 6]
```diff
@@ -57,6 +57,8 @@ static const char *RcsId = "$Id:  $";
 #include <helpers/PogoHelper.h>
 #include <Hamamatsu.h>
 #include <HamamatsuClass.h>
+#include <string> 
+#include <string.h>
```
Comment (soleil-ica) [inline specifics/Hamamatsu/Hamamatsu.cpp:61]: string.h should be removed I think

File: specifics/Hamamatsu/Hamamatsu.cpp [hunk 7]
```diff
@@ -627,6 +649,331 @@ void Hamamatsu::read_attr_hardware(vector<long> &attr_list)
...
+void Hamamatsu::read_channel1Polarity(Tango::Attribute &attr)
+{
+	DEBUG_STREAM << "Hamamatsu::read_channel1Polarity(Tango::Attribute &attr) entering... "<< endl;
+
+    try
+	{ 
+		*attr_channel1Polarity_read = (Tango::DevUShort)m_camera->getOutputTriggerPolarity(CHANNEL_1);
+		attr.set_value(attr_channel1Polarity_read);
+	}
+    catch(Tango::DevFailed & df)
```
Comment (soleil-ica) [inline specifics/Hamamatsu/Hamamatsu.cpp:668]: not properly aligned

File: specifics/Hamamatsu/Hamamatsu.cpp [hunk 8]
```diff
@@ -627,6 +649,331 @@ void Hamamatsu::read_attr_hardware(vector<long> &attr_list)
...
+	DEBUG_STREAM << "Hamamatsu::write_channel1Polarity(Tango::WAttribute &attr) entering... "<< endl;
+
+     try
+	{
+        attr.get_write_value(attr_channel1Polarity_write);
+        m_camera->setOutputTriggerPolarity(CHANNEL_1, (lima::Hamamatsu::Camera::Output_Trigger_Polarity)attr_channel1Polarity_write);
+
+		PropertyHelper::set_property(this, "MemorizedChannel1Polarity", attr_channel1Polarity_write);
+	}
+    catch(Tango::DevFailed & df)
```
Comment (soleil-ica) [inline specifics/Hamamatsu/Hamamatsu.cpp:696]: idem

**Review by soleil-ica — CHANGES_REQUESTED** [review]

**Review by soleil-ica — APPROVED** [review]

**Pull request merged**